monkey/monkey_island/cc/ui/src/components/ui-components/AdvancedMultiSelect.js

```diff
@@ -41,81 +32,168 @@ function getObjectFromRegistryByRef(refString, registry) {
   return resolveObjectPath(refArray, registry);
 }
 
-function getFullDefinitionByKey(refString, registry, itemKey) {
-  let fullArray = getFullDefinitionsFromRegistry(refString, registry);
-  return fullArray.filter(e => (e.enum[0] === itemKey))[0];
-}
-
-function setPaneInfo(refString, registry, itemKey, setPaneInfoFnc) {
-  let definitionObj = getFullDefinitionByKey(refString, registry, itemKey);
-  setPaneInfoFnc({title: definitionObj.title, content: definitionObj.info, link: definitionObj.link});
-}
-
 function getDefaultPaneParams(refString, registry) {
   let configSection = getObjectFromRegistryByRef(refString, registry);
   return ({title: configSection.title, content: configSection.description});
 }
 
-function AdvancedMultiSelect(props) {
-  const [masterCheckbox, setMasterCheckbox] = useState(true);
-  const {
-    schema,
-    id,
-    options,
-    value,
-    required,
-    disabled,
-    readonly,
-    multiple,
-    autofocus,
-    onChange,
-    registry
-  } = props;
-  const {enumOptions} = options;
-  const [infoPaneParams, setInfoPaneParams] = useState(getDefaultPaneParams(schema.items.$ref, registry));
-  getDefaultPaneParams(schema.items.$ref, registry);
-  const selectValue = cloneDeep(value);
-  return (
-    <div className={'advanced-multi-select'}>
-      <Card.Header>
-        <Button key={`${props.schema.title}-button`} value={value}
+class AdvancedMultiSelect extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      masterCheckboxState: this.getMasterCheckboxState(props.value),
+      infoPaneParams: getDefaultPaneParams(props.schema.items.$ref, props.registry)
+    };
+
+    this.onMasterCheckboxClick = this.onMasterCheckboxClick.bind(this);
+    this.onChildCheckboxClick = this.onChildCheckboxClick.bind(this);
```

Comment:
Boilerplate code, why not use arrow functions instead?

Reply:
According to https://reactjs.org/docs/handling-events.html
> The problem with [arrow function] syntax is that a different callback is created each time the LoggingButton renders. In most cases, this is fine. However, if this callback is passed as a prop to lower components, those components might do an extra re-rendering. We generally recommend binding in the constructor or using the class fields syntax, to avoid this sort of performance problem.

Reply:
Sorry, I got confused, I meant "public class fields syntax" (since it uses arrow, I though it was the same thing, just in definition :D)
```
onChildCheckboxClick = () => {
    // some js code
  }
```

Reply:
According to the docs, that syntax is still experimental.

> Please note that the syntax above is experimental and the syntax may change, or the proposal might not make it into the language.
https://reactjs.org/docs/react-without-es6.html#autobinding

It could be added inside the constructor as shown below, but I don't think it enhances readability. On the contrary, it makes the constructor artificially long.

```
constructor(props) {
  super(props);
  ...
  this.onChildCheckboxClick = () => {
    // some js code
  }
  ...
}
```

What if I add a `// TODO` to remind us to change to "Class Properties" syntax if it ever gets formally accepted into the language?

Reply:
It doesn't need to be in the constructor, this pattern is applied to the function definition (line 39) to autobind. It alters function definition slightly, but avoids explicit code lines for method binding.

Reply:
Adding TODO to refactor later is not worth it IMO

Reply:
> It doesn't need to be in the constructor, this pattern is applied to the function definition (line 39) to autobind.

Can you provide a link to an example? What I found indicated that this kind of syntax was still experimental and shouldn't be used in production code.

Reply:
It appears this syntax is currently a stage 3 proposal and has been implemented in Node 12. We also use it elsewhere throughout the code. If there are changes to the syntax at this point (unlikely), then we'll have to fix a bunch of these, not just this one. Given this information, I'll update the code to use class properties syntax.